### PR TITLE
Adding indexes

### PR DESCRIPTION
**Pull request by josenavas**
Adds the missing gui for adding indexes.

### CODE REVIEW

**Comment by josenavas** [comment]
@qiyun @wasade available for a review?

File: knimin/handlers/pm_library_prep.py [hunk 1]
```diff
@@ -54,4 +58,54 @@ def post(self):
...
+            plate['shotgun_plate_id'])['name']
+
+        contents = format_index_echo_pick_list(plate['shotgun_index'], volume)
+        file_name = "echo_index_PickList_%s_%s_%s.csv" % (
+            plate['shotgun_normalized_plate_id'], plate['name'],
+            date.today().strftime('%Y_%m_%d'))
+
+        self.set_header('Content-Type', 'application/octet-stream')
+        self.set_header('Content-Disposition',
+                        'attachment; filename=' + file_name)
```
Comment (qiyunzhu) [inline knimin/handlers/pm_library_prep.py]: Is there a reason here using `+` instead of `%`?

File: knimin/lib/format.py [hunk 2]
```diff
@@ -28,6 +32,62 @@ def format_epmotion_file(volumes, destination):
...
 
+def format_index_echo_pick_list(idx_layout, volume):
+    """Formats the contents of an echo indexing pick list file
+
+    Parameters
+    ----------
+    idx_layout: 2d numpy array of ints
+        The per well index information
+    vol : float
+        The v
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: Some text is mssing here?

File: knimin/templates/pm_metagenomics_library_prep.html [hunk 3]
```diff
@@ -1,8 +1,84 @@
...
+    var formData = new FormData($('#pm-form')[0]);
+
+    $.ajax({
+      url: '/pm_library_prep/metagenomics/',
+      data: formData,
+      processData: false,
+      contentType: false,
+      type: 'POST',
+      success: function(data){
+        var win = window.open('/pm_library_prep/metagenomics/echo/?plate_id={{plate['shotgun_normalized_plate_id']}}&volume=' + formData.get('volume'), '_blank');
```
Comment (qiyunzhu) [inline knimin/templates/pm_metagenomics_library_prep.html]: The variable `win` does not seem to be used anywhere. Is it necessary to declare it explicitly?

File: knimin/templates/pm_metagenomics_library_prep.html [hunk 4]
```diff
@@ -1,8 +1,84 @@
...
+      <div class='col-xs-4 form-group'>
+        <label>Choose library prep kit</label>
+        <select class="form-control" name="kit" required>
+          {% for lk in library_kits %}
+            <option value='{{lk['name']}}'>{{lk['name']}}</option>
+          {% end %}
+        </select>
+      </div>
+    </div>
+    <!-- Second row -->
```
Comment (qiyunzhu) [inline knimin/templates/pm_metagenomics_library_prep.html]: Is there also `first row` somewhere? (line #34?)

File: knimin/lib/format.py [hunk 5]
```diff
@@ -28,6 +32,62 @@ def format_epmotion_file(volumes, destination):
...
+
+    i5contents = []
+    i7contents = []
+    vol = "%.3f" % volume
+
+    for i in range(rows):
+        for j in range(cols):
+            if idx_layout[i, j] != 0:
+                idx_info = db.get_shotgun_index_information(idx_layout[i, j])
+                dest = "%s%s" % (chr(ord('A') + i), j + 1)
```
Comment (wasade) [inline knimin/lib/format.py]: This is going to be prone to an easy error, and is repeated. Can this be decomposed to:

```python
def well(i, j):
    return (chr(ord('A') + i), j + 1)
```

File: knimin/lib/shotgun.py [hunk 6]
```diff
@@ -202,3 +204,33 @@ def compute_shotgun_normalization_values(plate_layout, input_vol, input_dna):
...
+    samples_vol = plate['plate_normalization_sample']
+    num_samples = (samples_vol > 0.0001).sum()
+
+    indexes = db.generate_i5_i7_indexes(idx_tech, num_samples)
+    rows, cols = samples_vol.shape
+    barcode_layout = np.zeros((rows, cols), dtype=np.int)
+    idx = 0
+    for i in range(rows):
+        for j in range(cols):
+            if samples_vol[i, j] > 0.0001:
```
Comment (wasade) [inline knimin/lib/shotgun.py]: > 0?

File: knimin/lib/format.py (hunk 2 above)
Comment (wasade) [inline knimin/lib/format.py]: done

File: knimin/lib/format.py (hunk 5 above)
Comment (wasade) [inline knimin/lib/format.py]: done

File: knimin/templates/pm_metagenomics_library_prep.html (hunk 3 above)
Comment (wasade) [inline knimin/templates/pm_metagenomics_library_prep.html]: i believe it is necessary but i'm not entirely sure.

File: knimin/templates/pm_metagenomics_library_prep.html (hunk 4 above)
Comment (wasade) [inline knimin/templates/pm_metagenomics_library_prep.html]: added

**Comment by wasade** [comment]
ugh. local changes which were not supposed to get committed did. rolling those back.

**Comment by wasade** [comment]
i did not `git add` those files. wtf.

File: knimin/lib/shotgun.py [hunk 7]
```diff
@@ -202,3 +204,35 @@ def compute_shotgun_normalization_values(plate_layout, input_vol, input_dna):
...
     return vol_sample, vol_water
+
+
+def prepare_shotgun_libraries(plate_id, email, mosquito, kit, aliquot,
+                              idx_tech):
+    """
+    Parameters
+    ----------
+    plate_id : int
+        The
```
Comment (mortonjt) [inline knimin/lib/shotgun.py]: `The`?

**Pull request merged**